### PR TITLE
Add str_studio cadence_16 keyboard

### PR DESCRIPTION
**Pull request by str-studio**
## Description

Add str_studio cadence_16 keyboard

## Types of Changes

- [ ] Core
- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement/optimization
- [x] Keyboard (addition or update)
- [ ] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Issues Fixed or Closed by This PR

* 

## Checklist

- [x] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [x] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [x] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

File: keyboards/str_studio/cadence_16/cadence_16.c
Comment (fauxpark) [inline keyboards/str_studio/cadence_16/cadence_16.c:1]: This file can be deleted.

File: keyboards/str_studio/cadence_16/config.h [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+
+/* key matrix size */
+#define MATRIX_ROWS 5
+#define MATRIX_COLS 4
+
+#define MATRIX_ROW_PINS { F0, F6, F5, F4, F1 }
+#define MATRIX_COL_PINS { B6, B5, B4, D7 } // A virtual pin (NO_PIN) is needed for the encoder key matrix in via.
+
+#define DIODE_DIRECTION COL2ROW
+
```
Comment (fauxpark) [inline keyboards/str_studio/cadence_16/config.h:27]:
```suggestion
```

File: keyboards/str_studio/cadence_16/info.json [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+    "usb": {
+        "vid": "0x5712",
+        "pid": "0x0001",
+        "device_version": "0.0.1"
+    },
+    "encoder": {
+        "rotary": [
+            {"pin_a": "D2", "pin_b": "D3", "resolution": 2}
+        ]
+    },
```
Comment (fauxpark) [inline keyboards/str_studio/cadence_16/info.json:13]:
```suggestion
    },
    "matrix_pins": {
        "cols" ["B6", "B5", "B4", "D7"],
        "rows": ["F0", "F6", "F5", "F4", "F1"]
    },
    "diode_direction": "COL2ROW",
```

File: keyboards/str_studio/cadence_16/keyboard_layout-default.json
Comment (fauxpark) [inline keyboards/str_studio/cadence_16/keyboard_layout-default.json:1]: Please remove this file.

File: keyboards/str_studio/cadence_16/cadence_16.h
Comment (fauxpark) [inline keyboards/str_studio/cadence_16/cadence_16.h:1]: This file can be deleted.

File: keyboards/str_studio/cadence_16/keyboard_layout_with_encoder.json
Comment (fauxpark) [inline keyboards/str_studio/cadence_16/keyboard_layout_with_encoder.json:1]: Please remove this file.

File: keyboards/str_studio/cadence_16/keymaps/default/config.h
Comment (fauxpark) [inline keyboards/str_studio/cadence_16/keymaps/default/config.h:1]: This file can be deleted.

File: keyboards/str_studio/cadence_16/keymaps/default/keymap.c [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+        if (clockwise) {
+            tap_code(KC_PGDN);
+        } else {
+            tap_code(KC_PGUP);
+        }
+    }
+    return true;
+}
+
+#endif
```
Comment (fauxpark) [inline keyboards/str_studio/cadence_16/keymaps/default/keymap.c:104]: This should use Encoder Map instead.

File: keyboards/str_studio/cadence_16/keymaps/default/keymap.c [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include QMK_KEYBOARD_H
+
+#include "oled.c"
```
Comment (fauxpark) [inline keyboards/str_studio/cadence_16/keymaps/default/keymap.c:19]: Do not include .c files in other .c files.

File: keyboards/str_studio/cadence_16/keymaps/default/readme.md
Comment (fauxpark) [inline keyboards/str_studio/cadence_16/keymaps/default/readme.md:1]: This file can be deleted.

File: keyboards/str_studio/cadence_16/keymaps/str-vial/config.h
Comment (fauxpark) [inline keyboards/str_studio/cadence_16/keymaps/str-vial/config.h:1]: This entire VIAL keymap must be deleted.

File: keyboards/str_studio/cadence_16/keymaps/via/config.h [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+ */
+
+#pragma once
+
+
+// Disabled to save space
+#define NO_ACTION_ONESHOT  // 332
+#define NO_ACTION_MACRO    // 0
+#define NO_ACTION_FUNCTION // 0
+#define DISABLE_LEADER     // 0
```
Comment (fauxpark) [inline keyboards/str_studio/cadence_16/keymaps/via/config.h:24]:
```suggestion
```

File: keyboards/str_studio/cadence_16/keymaps/via/encoder.c
Comment (fauxpark) [inline keyboards/str_studio/cadence_16/keymaps/via/encoder.c:1]: VIA supports Encoder Map; there is no need for this. Please remove it.

File: keyboards/str_studio/cadence_16/rules.mk [hunk 6]
```diff
@@ -0,0 +1,22 @@
+# MCU name
+MCU = atmega32u4
+
```
Comment (fauxpark) [inline keyboards/str_studio/cadence_16/rules.mk:3]:
```suggestion
```

**Review by fauxpark — CHANGES_REQUESTED** [review]

File: keyboards/str_studio/cadence_16/cadence_16.h
Comment (fauxpark) [inline keyboards/str_studio/cadence_16/cadence_16.h:1]: This file should be deleted. You already have matrix positions defined in info.json. If they are incorrect, you need to fix them there instead.

File: keyboards/str_studio/cadence_16/info.json
Comment (fauxpark) [inline keyboards/str_studio/cadence_16/info.json:1]: Please run `qmk format-json -i` on this file.

**Review by fauxpark — CHANGES_REQUESTED** [review]

File: keyboards/str_studio/cadence_16/info.json
Comment (str-studio) [reply keyboards/str_studio/cadence_16/info.json:1]: Thanks for the suggestions!

File: keyboards/str_studio/cadence_16/keymaps/default/keymap.c [hunk 7]
```diff
@@ -0,0 +1,134 @@
...
+    /* _M_XYZ = Mac Os, _W_XYZ = Win/Linux */
+    _NUMBER,
+    _FN
+};
+
+enum custom_keycodes {
+    PLACEHOLDER = SAFE_RANGE,
+    CYCLE
+};
+
```
Comment (zvecr) [inline keyboards/str_studio/cadence_16/keymaps/default/keymap.c:29]:
```suggestion
```

File: keyboards/str_studio/cadence_16/keymaps/default/keymap.c [hunk 8]
```diff
@@ -0,0 +1,134 @@
...
+   KC_LEFT,     KC_DOWN,    KC_RGHT,        KC_KP_ASTERISK,
+   QK_BOOT,     KC_HOME,    KC_PAGE_UP,     KC_KP_ENTER,
+   MO(_FN),     KC_END,     KC_PAGE_DOWN
+)
+
+};
+
+// ================== Encoder ==================
+#if defined(ENCODER_MAP_ENABLE)
+const uint16_t PROGMEM encoder_map[][NUM_ENCODERS][2] = {
```
Comment (zvecr) [inline keyboards/str_studio/cadence_16/keymaps/default/keymap.c:81]:
```suggestion
const uint16_t PROGMEM encoder_map[][NUM_ENCODERS][NUM_DIRECTIONS] = {
```

File: keyboards/str_studio/cadence_16/keymaps/default/oled.c
Comment (zvecr) [inline keyboards/str_studio/cadence_16/keymaps/default/oled.c:1]: This file is not being compiled, and should be removed.

File: keyboards/str_studio/cadence_16/keymaps/default/keymap.c [hunk 9]
```diff
@@ -0,0 +1,134 @@
...
+#endif
+
+
+// ================== OLED ==================
+
+#ifdef OLED_ENABLE
+
+/* advanced settings */
+#    define ANIM_FRAME_DURATION 200  // how long each frame lasts in ms
+#    define ANIM_SIZE           128   // number of bytes in array. If you change sprites, minimize for adequate firmware size. max is 1024
```
Comment (zvecr) [inline keyboards/str_studio/cadence_16/keymaps/default/keymap.c:95]:
```suggestion
```

File: keyboards/str_studio/cadence_16/keymaps/default/keymap.c [hunk 10]
```diff
@@ -0,0 +1,134 @@
...
+        case 0: // _NUMBER
+            oled_write_P(PSTR("Num\n"), false);
+            break;
+        case 1: // _FN
+            oled_write_P(PSTR("Fn\n"), false);
+            break;
+        default:
+            oled_write_P(PSTR("Undef"), false);
+    }
+    // oled_write_P(PSTR("\n\n"), false);
```
Comment (zvecr) [inline keyboards/str_studio/cadence_16/keymaps/default/keymap.c:113]:
```suggestion
```

File: keyboards/str_studio/cadence_16/keymaps/default/keymap.c [hunk 11]
```diff
@@ -0,0 +1,134 @@
...
+#ifdef OLED_ENABLE
+
+/* advanced settings */
+#    define ANIM_FRAME_DURATION 200  // how long each frame lasts in ms
+#    define ANIM_SIZE           128   // number of bytes in array. If you change sprites, minimize for adequate firmware size. max is 1024
+
+static void print_status_narrow(void) {
+    // Print current mode
+    // oled_write_P(PSTR("\n\n"), false);
+
```
Comment (zvecr) [inline keyboards/str_studio/cadence_16/keymaps/default/keymap.c:100]:
```suggestion
```

File: keyboards/str_studio/cadence_16/keymaps/default/keymap.c [hunk 12]
```diff
@@ -0,0 +1,134 @@
...
+    }
+    return rotation;
+}
+
+bool oled_task_user(void) {
+    if (is_keyboard_master()) {
+        print_status_narrow();
+    } else {
+        print_status_narrow();
+    }
```
Comment (zvecr) [inline keyboards/str_studio/cadence_16/keymaps/default/keymap.c:130]:
```suggestion
    print_status_narrow();
```

File: keyboards/str_studio/cadence_16/keymaps/default/rules.mk [hunk 13]
```diff
@@ -0,0 +1,5 @@
+OLED_ENABLE = yes
+OLED_DRIVER = ssd1306
+
+ENCODER_MAP_ENABLE = yes    # Enable Encoder Map
+ENCODER_ENABLE = yes
```
Comment (zvecr) [inline keyboards/str_studio/cadence_16/keymaps/default/rules.mk:5]:
```suggestion
OLED_ENABLE = yes
ENCODER_MAP_ENABLE = yes
```

File: keyboards/str_studio/cadence_16/readme.md [hunk 14]
```diff
@@ -0,0 +1,23 @@
...
+
+    qmk compile -kb str_studio/cadence_16 -km default
+
+Flashing the firmware using [QMK Toolbox](https://github.com/qmk/qmk_toolbox)
+
+See the [build environment setup](https://docs.qmk.fm/#/getting_started_build_tools) and the [make instructions](https://docs.qmk.fm/#/getting_started_make_guide) for more information. Brand new to QMK? Start with our [Complete Newbs Guide](https://docs.qmk.fm/#/newbs).
+
+## Bootloader
+
+Enter the bootloader in 3 ways:
```
Comment (zvecr) [inline keyboards/str_studio/cadence_16/readme.md:20]:
```suggestion
Enter the bootloader in 2 ways:
```

File: keyboards/str_studio/cadence_16/rules.mk [hunk 15]
```diff
@@ -0,0 +1,19 @@
...
+BACKLIGHT_ENABLE = no       # Enable keyboard backlight functionality
+RGBLIGHT_ENABLE = no        # Enable keyboard RGB underglow
+AUDIO_ENABLE = no           # Audio output
+
+WPM_ENABLE = yes			# Enable WPM counter
+ENCODER_ENABLE = yes		# Enable Encoder
+
+LTO_ENABLE = yes			# Enable Link Time Optimization
+MIDI_ENABLE = yes
```
Comment (zvecr) [inline keyboards/str_studio/cadence_16/rules.mk:19]:
```suggestion
# This file intentionally left blank
```

File: keyboards/str_studio/cadence_16/info.json [hunk 16]
```diff
@@ -0,0 +1,43 @@
+{
+    "manufacturer": "str studio",
+    "keyboard_name": "cadence-16",
+    "bootloader": "atmel-dfu",
+    "diode_direction": "ROW2COL",
+    "encoder": {
+        "rotary": [
+            {"pin_a": "D2", "pin_b": "D3", "resolution": 2}
+        ]
+    },
```
Comment (zvecr) [inline keyboards/str_studio/cadence_16/info.json:10]:
```suggestion
    },
    "features": {
        "bootmagic": true,
        "extrakey": true,
        "mousekey": true,
        "nkro": true,
        "encoder": true
    },
```

**Review by zvecr — CHANGES_REQUESTED** [review]

File: keyboards/str_studio/cadence_16/info.json [hunk 17]
```diff
@@ -0,0 +1,50 @@
...
+    "matrix_pins": {
+        "cols": ["B6", "B5", "B4", "D7"],
+        "rows": ["F0", "F6", "F5", "F4", "F1"]
+    },
+    "processor": "atmega32u4",
+    "usb": {
+        "device_version": "0.0.1",
+        "pid": "0x0001",
+        "vid": "0x5712"
+    },
```
Comment (fauxpark) [inline keyboards/str_studio/cadence_16/info.json:27]:
```suggestion
    },
    "qmk": {
        "tap_keycode_delay": 10
    },
```

File: keyboards/str_studio/cadence_16/config.h [hunk 18]
```diff
@@ -0,0 +1,23 @@
...
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#pragma once
+
+#define TAP_CODE_DELAY  10
+
```
Comment (fauxpark) [inline keyboards/str_studio/cadence_16/config.h:20]:
```suggestion
```

**Review by fauxpark — CHANGES_REQUESTED** [review]

**Comment by str-studio** [comment]
Hey @fauxpark @zvecr, really appreciate for your suggestions. I was wondering if you could help review the code again? So that it can be able merge into the repo. 🙏

**Review by fauxpark — APPROVED** [review]

**Comment by str-studio** [comment]
Hello @zvecr , 

I hope you're doing well. Just circling back to the review request that's been waiting in the pipeline for a few days. I totally understand how packed your schedule must be, and I appreciate your time in reviewing the code.

Following your valuable advice, I've updated the code accordingly. If you could find a moment to approve this pull request, it would be great. It'll allow us to integrate the keyboard and keymap files into the official firmware, giving users the ability to utilize and personalize them as per their needs.

Thanks a lot in advance for your help with this. Really appreciate it!

**Comment by str-studio** [comment]
@zvecr Hi, Joel. Any updates for this MR? It has been idling for several weeks, feel free to lmk if anything need to be fixed? I'd appreciate it

File: keyboards/str_studio/cadence_16/keymaps/default/keymap.c [hunk 19]
```diff
@@ -0,0 +1,116 @@
...
+    [_NUMBER] =   { ENCODER_CCW_CW(KC_VOLD, KC_VOLU)  },
+    [_FN]     =   { ENCODER_CCW_CW(KC_PGDN, KC_PGUP)  }
+    //                  Encoder 1              
+};  
+#endif
+
+
+// ================== OLED ==================
+
+#ifdef OLED_ENABLE
```
Comment (drashna) [inline keyboards/str_studio/cadence_16/keymaps/default/keymap.c:85]: Ideally, the oled code should be in the keyboard's c file, so there is a default behavior (eg when it's compiled by the configurator, etc). And use the kb function.

**Pull request closed without merging**